dsniff.8
.TH DSNIFF 8
.ad
.fi
.SH NAME
dsniff
\-
password sniffer
.SH SYNOPSIS
.na
.nf
.fi
\fBdsniff\fR [\fB-c\fR] [\fB-d\fR] [\fB-m\fR] [\fB-n\fR] [\fB-i
\fIinterface\fR | \fB-p \fIpcapfile\fR] [\fB-s \fIsnaplen\fR] [\fB-f \fIservices\fR]
[\fB-t \fItrigger[,...]\fR]]
[\fB-r\fR|\fB-w\fR \fIsavefile\fR] [\fIexpression\fR]
.SH DESCRIPTION
.ad
.fi
\fBdsniff\fR is a password sniffer which handles FTP, Telnet, SMTP,
HTTP, POP, poppass, NNTP, IMAP, SNMP, LDAP, Rlogin, RIP, OSPF, PPTP
MS-CHAP, NFS, VRRP, YP/NIS, SOCKS, X11, CVS, IRC, AIM, ICQ, Napster,
PostgreSQL, Meeting Maker, Citrix ICA, Symantec pcAnywhere, NAI
Sniffer, Microsoft SMB, Oracle SQL*Net, Sybase and Microsoft SQL
protocols.
.LP
\fBdsniff\fR automatically detects and minimally parses each
application protocol, only saving the interesting bits, and uses
Berkeley DB as its output file format, only logging unique
authentication attempts. Full TCP/IP reassembly is provided by
libnids(3).
.LP
I wrote \fBdsniff\fR with honest intentions - to audit my own network,
and to demonstrate the insecurity of cleartext network protocols.
Please do not abuse this software.
.SH OPTIONS
.IP \fB-c\fR
Perform half-duplex TCP stream reassembly, to handle asymmetrically
routed traffic (such as when using arpspoof(8) to intercept client
traffic bound for the local gateway).
.IP \fB-d\fR
Enable debugging mode.
.IP \fB-m\fR
Enable automatic protocol detection.
.IP \fB-n\fR
Do not resolve IP addresses to hostnames.
.IP "\fB-i \fIinterface\fR"
Specify the interface to listen on.
.IP "\fB-p \fIpcapfile\fR"
Rather than processing the contents of packets observed upon the network 
process the given PCAP capture file.
.IP "\fB-s \fIsnaplen\fR"
Analyze at most the first \fIsnaplen\fR bytes of each TCP connection,
rather than the default of 1024.
.IP "\fB-f \fIservices\fR"
Load triggers from a \fIservices\fR file.
.IP "\fB -t \fItrigger\fR[,...]"
Load triggers from a comma-separated list, specified as
\fIport\fR/\fIproto\fR=\fIservice\fR (e.g. 80/tcp=http).
.IP "\fB-r \fIsavefile\fR"
Read sniffed sessions from a \fIsavefile\fR created with the \fB-w\fR
option.
.IP "\fB-w \fIfile\fR"
Write sniffed sessions to \fIsavefile\fR rather than parsing and
printing them out.
.IP "\fIexpression\fR"
Specify a tcpdump(8) filter expression to select traffic to sniff.
.LP
On a hangup signal \fBdsniff\fR will dump its current trigger table to
\fIdsniff.services\fR.
.SH FILES
.IP \fI/usr/share/dsniff/dsniff.services\fR
Default trigger table
.IP \fI/usr/share/dsniff/dsniff.magic\fR
Network protocol magic
.SH "SEE ALSO"
arpspoof(8), libnids(3), services(5), magic(5)
.SH AUTHOR
.na
.nf
Dug Song <dugsong@monkey.org>
.SH BUGS
\fBdsniff\fR's automatic protocol detection feature is based on the
classic file(1) command by Ian Darwin, and shares its historical
limitations and bugs.


CHANGES
$Id: CHANGES,v 1.56 2001/03/19 06:53:47 dugsong Exp $

- Add support for ICMP frag-needed to tcpnice.

- New sshow program:
  http://www.openwall.com/advisories/OW-003-ssh-traffic-analysis.txt

- Add support for HTTP absolute URIs to webmitm and urlsnarf.

- Add support for non-transparent proxying to webmitm, as requested
  by Juergen Schmidt <ju@ct.heise.de>.

- Fix TDS decode for MSSQL 7.x, from Paul van Maaren
  <P.v.Maaren@reseau.nl>.

- Kludge around stupid Redhat Linux 6, 7 build env inconsistencies.

- Fix dumb endianness bug in dnsspoof.


v2.3 Sun Dec 17 11:35:38 EST 2000

- Add VRRP parsing to dsniff, from Eric Jackson <shinobi@monkey.org>.

- Require pcap filter argument for tcpkill, tcpnice.

- Add Microsoft PPTP MS-CHAP (v1, v2) parsing to dsniff, based on
  anger.c by Aleph One <aleph1@securityfocus.com>.

- Fix pcAnywhere 7, 9.x parsing in dsniff.

- Add -t trigger[,...] flag to dsniff, to specify individual triggers
  on the command line.

- Convert most everything to use new buf interface.

- New programs: dnsspoof, msgsnarf, sshmitm, webmitm.

- Fix inverted regex matching in *snarf programs.

- Consistent arpspoof, macof, tcpnice, tcpkill output.

- Rename arpredirect to arpspoof (maintain consistent *sniff, *snarf,
  *spoof, *spy nomenclature).

- Consistent pcap filter argument to dsniff, *snarf programs.

- Add trigger for Checkpoint Firewall-1 Session Authentication Agent
  (261/tcp), as suggested by Joe Segreti <seg@clark.net>.

- Add SMTP parsing to dsniff, as requested by Denis Ducamp
  <Denis.Ducamp@hsc.fr>.

- Add rexec and RPC ypserv parsing to dsniff, as requested by 
  Oliver Friedrichs <of@securityfocus.com>.

- Add HTTP proxy auth parsing back to dsniff, it got lost in the
  shuffle. Reported by Denis Ducamp <Denis.Ducamp@hsc.fr>.

- Add NNTPv2 and other AUTHINFO extensions to dsniff.


v2.2 Wed Jun 14 00:58:37 EDT 2000

- Rewrite HTTP decoding in dsniff, adding support for QUERY_STRING and
  x-www-form-urlencoded parsing (various CGI authentication schemes).
  
- Alpha support (libnids and libnet still need to be fixed).

- Fix arp discovery in arpredirect on Linux.

- Add -m flag to enable automatic protocol detection in dsniff,
  based on the classic file(1) command by Ian Darwin.

- Add TDS (Sybase, Microsoft SQL Server) parsing to dsniff.

- Clean up RPC decodes, TCP half-duplex reassembly in dsniff.

- New filesnarf program.

- Add regular expression matching to mailsnarf.

- Add POP support to mailsnarf.


v2.1 Thu May 18 16:18:35 EDT 2000

- Add -c flag to specify half-duplex TCP stream reassembly in dsniff
  (better support for sniffing off switched ports using arpredirect).

- Fix > 24 char Meeting Maker passwd parsing in dsniff.

- Fix OSPF parsing in dsniff (don't truncate first two chars),
  as reported by Felix Contreras <cfelix@fisiologia.com>.

- Fix webspy URL ignoring, as reported by Interrupt <mike@eEye.com>.


v2.0 Tue May 16 13:11:22 EDT 2000

- Major dsniff rewrite, since ppl are actually reading this code. :-)

- Add configurable decode triggers to dsniff.

- Add dsniff debugging functions, split out decode routines.

- Add yppasswd parsing to dsniff.

- Rewrite dsniff RPC framework, portmap and NFS mountd decodes.

- Make dsniff savefile format portable.

- Remove findgw - to be subsumed by dsquat package.

- Add PostgreSQL parsing to dsniff.

- Add Meeting Maker parsing to dsniff.

- Add poppass parsing to dsniff.

- Add RIP, OSPF parsing to dsniff.

- Fix RSET handling in mailsnarf (from Martin Fredriksson <martin@crt.se>).


v1.8 Sun Apr  9 23:59:46 EDT 2000

- Add SOCKS parsing to dsniff.

- Add pcAnywhere parsing to dsniff.

- Fix SMB parsing in dsniff.

- Add IRC parsing to dsniff.

- Add NAI Sniffer parsing to dsniff (from Anonymous).


v1.7 Mon Mar 27 16:19:32 EST 2000

- Add -s flag to specify snaplen to dsniff.

- Support systems without <libgen.h> or dirname().

- Add Microsoft SMB parsing to dsniff.

- Add Citrix ICA parsing to dsniff.

- Add LDAP parsing to dsniff.

- Fix Berkeley mbox format again (\n, not \r\n).

- Fix null URI dereference in urlsnarf.

- Add Oracle SQL*Net (v2, Net8) parsing to dsniff.

- Catch data left on connection close in mailsnarf, urlsnarf, webspy.


v1.6 Sun Mar 12 16:25:09 EST 2000

- Support non-glibc Linux systems missing ether_ntoa().

- Unique HTTP auth info by URI dirname in dsniff.

- Add Napster parsing to dsniff.

- Don't rely on /etc/services for dsniff.

- Add AIM, ICQ (v2, v5) parsing to dsniff.

- Add CVS pserver parsing to dsniff.

- Skip IMAP command tag in dsniff.


v1.5 Tue Feb 15 23:22:25 EST 2000

- Fix HTTP proxy support in urlsnarf (from <felix@convergence.de>).

- Fix HTTP proxy support in dsniff (from <Alain.Thivillon@hsc.fr>).

- Proper manpages for all programs.

- Strip binary nulls in telnet input, in dsniff (doh!).


v1.4 Thu Jan 27 12:08:41 EST 2000

- Add verbose flag (-v) to tcpkill, tcpnice.

- Add NNTP parsing to dsniff (from Felix von Leitner <felix@convergence.de>).

- Fix mailsniff mbox formatting of ^From in message body.

- Add HTTP proxy support in dsniff, urlsnarf, webspy.

- Fix getopt() usage to be POSIX compliant (s/EOF/-1/).

- New tcpnice program.


v1.3 Fri Jan 21 02:47:37 EST 2000

- Ported to Solaris (along with libnids :-)

- Add Berkeley db(3) output file format to dsniff, as well as
  restricting logging to unique auth info.

- New tcpkill program.

- New lame dsniff(8) manpage.

- Add DNS lookups (and -n flag to disable) in dsniff, urlsnarf.

- Add HTTP Basic Authentication, Referer, User-Agent logging to urlsnarf.

- Improve RPC message parsing in dsniff.

- Improve SMTP parsing in mailsnarf.

- Improve HTTP 1.x parsing in dsniff, urlsnarf, webspy.

- Fix IMAP, Rlogin, Telnet option parsing in dsniff (broke them in 1.2).

- Add X11 MIT-MAGIC-COOKIE parsing to dsniff.

- Don't forget to decode POP SASL username in dsniff (doh!).


v1.2 Sat Jan  8 22:36:42 EST 2000

- Ported to FreeBSD (but not tested!).

- Add GNU autoconf support.


dsniff.c
/*
 * dsniff.c
 *
 * Password sniffer, because DrHoney wanted one.
 *
 * This is intended for demonstration purposes and educational use only.
 *
 * Copyright (c) 2000 Dug Song <dugsong@monkey.org>
 *
 * $Id: dsniff.c,v 1.69 2001/03/15 08:33:03 dugsong Exp $
 */

#include "config.h"

#include <sys/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <err.h>
#include <libnet.h>
#include <nids.h>
#include <pcap.h>

#include "options.h"
#include "pathnames.h"
#include "pcaputil.h"
#include "trigger.h"
#include "record.h"
#include "version.h"

#define MAX_LINES	6
#define MIN_SNAPLEN	1024

int	Opt_client = 0;
int	Opt_debug = 0;
u_short	Opt_dns = 1;
int	Opt_magic = 0;
int	Opt_read = 0;
int	Opt_write = 0;
int	Opt_snaplen = MIN_SNAPLEN;
int	Opt_lines = MAX_LINES;

static void
usage(void)
{
	fprintf(stderr, "Version: " VERSION "\n"
		"Usage: dsniff [-cdmn] [-i interface | -p pcapfile] [-s snaplen]\n"
		"              [-f services] [-t trigger[,...]] [-r|-w savefile]\n"
		"              [expression]\n");
	exit(1);
}

static void
sig_hup(int sig)
{
	trigger_dump();
}

static void
sig_die(int sig)
{
	record_close();
	exit(0);
}

static void
null_syslog(int type, int errnum, struct ip *iph, void *data)
{
}


static int get_all_ifaces(struct ifreq **, int *);
static unsigned int get_addr_from_ifreq(struct ifreq *);

int all_local_ipaddrs_chksum_disable()
{
	struct ifreq *ifaces;
	int ifaces_count;
	int i, ind = 0;
	struct nids_chksum_ctl *ctlp;
	unsigned int tmp;

	if (!get_all_ifaces(&ifaces, &ifaces_count))
		return -1;
	ctlp =
	    (struct nids_chksum_ctl *) malloc(ifaces_count *
					      sizeof(struct
						     nids_chksum_ctl));
	if (!ctlp)
		return -1;
	for (i = 0; i < ifaces_count; i++) {
		tmp = get_addr_from_ifreq(ifaces + i);
		if (tmp) {
			ctlp[ind].netaddr = tmp;
			ctlp[ind].mask = inet_addr("255.255.255.255");
			ctlp[ind].action = NIDS_DONT_CHKSUM;
			ind++;
		}
	}
	free(ifaces);
	nids_register_chksum_ctl(ctlp, ind);
}

/* helper functions for Example 2 */
unsigned int get_addr_from_ifreq(struct ifreq *iface)
{
	if (iface->ifr_addr.sa_family == AF_INET)
		return ((struct sockaddr_in *) &(iface->ifr_addr))->
		    sin_addr.s_addr;
	return 0;
}

static int get_all_ifaces(struct ifreq **ifaces, int *count)
{
	int ifaces_size = 8 * sizeof(struct ifreq);
	struct ifconf param;
	int sock;
	unsigned int i;

	*ifaces = malloc(ifaces_size);
	sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_IP);
	if (sock <= 0)
		return 0;
	for (;;) {
		param.ifc_len = ifaces_size;
		param.ifc_req = *ifaces;
		if (ioctl(sock, SIOCGIFCONF, &param))
			goto err;
		if (param.ifc_len < ifaces_size)
			break;
		free(*ifaces);
		ifaces_size *= 2;
		ifaces = malloc(ifaces_size);
	}
	*count = param.ifc_len / sizeof(struct ifreq);
	close(sock);
	return 1;
      err:
	close(sock);
	return 0;
}



int
main(int argc, char *argv[])
{
	extern char *optarg;
	extern int optind;
	char *services, *savefile, *triggers;
	int c;

	services = savefile = triggers = NULL;
	
	while ((c = getopt(argc, argv, "cdf:i:mnp:r:s:t:w:h?V")) != -1) {
		switch (c) {
		case 'c':
			Opt_client = 1;
			break;
		case 'd':
			Opt_debug++;
			break;
		case 'f':
			services = optarg;
			break;
		case 'i':
			nids_params.device = optarg;
			break;
		case 'm':
			Opt_magic = 1;
			break;
		case 'n':
			Opt_dns = 0;
			break;
		case 'p':
			nids_params.filename = optarg;
			break;
		case 'r':
			Opt_read = 1;
			savefile = optarg;
			break;
		case 's':
			if ((Opt_snaplen = atoi(optarg)) == 0)
				usage();
			break;
		case 't':
			triggers = optarg;
			break;
		case 'w':
			Opt_write = 1;
			savefile = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;
	
	if (Opt_read && Opt_write)
		usage();
	
	if (!record_init(savefile))
		err(1, "record_init");
	
	signal(SIGHUP, sig_hup);
	signal(SIGINT, sig_die);
	signal(SIGTERM, sig_die);
	
	if (Opt_read) {
		record_dump();
		record_close();
		exit(0);
	}

	if (argc != 0)
		nids_params.pcap_filter = copy_argv(argv);
	nids_params.scan_num_hosts = 0;
	nids_params.syslog = null_syslog;
	
	if (!nids_init()) {
		record_close();
		errx(1, "nids_init: %s", nids_errbuf);
	}
	if (Opt_magic) {
		trigger_init_magic(DSNIFF_LIBDIR DSNIFF_MAGIC);
	}
	if (triggers) {
		trigger_init_list(triggers);
	}
	if (services == NULL) {
		services = DSNIFF_LIBDIR DSNIFF_SERVICES;
	}
	trigger_init_services(services);
	
	nids_register_ip(trigger_ip);
	nids_register_ip(trigger_udp);
		
	if (Opt_client) {
		nids_register_ip(trigger_tcp_raw);
		signal(SIGALRM, trigger_tcp_raw_timeout);
		alarm(TRIGGER_TCP_RAW_TIMEOUT);
	}
	else nids_register_tcp(trigger_tcp);
	
	if (nids_params.pcap_filter != NULL) {
		if (nids_params.filename == NULL) {
			warnx("listening on %s [%s]", nids_params.device,
		        nids_params.pcap_filter);
		}
		else {
			warnx("using %s [%s]", nids_params.filename,
		        nids_params.pcap_filter);
		}
	}
	else {
		if (nids_params.filename == NULL) {
			warnx("listening on %s", nids_params.device);
		}
		else {
			warnx("using %s", nids_params.filename);
		}
	}

	all_local_ipaddrs_chksum_disable();
	
	nids_run();
	
	/* NOTREACHED */
	
	exit(0);
}
